compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1567,6 +1575,97 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+    ) -> Option<&'b hir::Expr<'b>> {
+        // When `expr` is `x` in something like `let x = foo.clone(); x`, need to recurse up to
+        // get `foo` and `clone`.
+        match expr.kind {
+            hir::ExprKind::Path(hir::QPath::Resolved(
+                None,
+                hir::Path { segments: [_], res: crate::Res::Local(binding), .. },
+            )) => {
+                let Some(hir::Node::Pat(hir::Pat { hir_id, .. })) = self.tcx.hir().find(*binding)
+                else {
+                    return None;
+                };
+                let Some(parent) = self.tcx.hir().find(self.tcx.hir().parent_id(*hir_id)) else {
+                    return None;
+                };
+
+                match parent {
+                    // foo.clone()
+                    hir::Node::Local(hir::Local { init: Some(init), .. }) => {
+                        return self.note_type_is_not_clone_inner_expr(init);
+                    }
+                    // When `expr` is more complex like a tuple
+                    hir::Node::Pat(hir::Pat { hir_id: pat_hir_id, kind: pat_kind, .. }) => {
+                        let Some(hir::Node::Local(hir::Local { init: Some(init), .. })) =
+                                self.tcx.hir().find(self.tcx.hir().parent_id(*pat_hir_id)) else {
+                            return None;
+                        };
+
+                        if let hir::PatKind::Tuple(pats, ..) = pat_kind {
+                            let find_init = |init_tup: &'b [Expr<'b>]| {
+                                pats.iter().enumerate().filter(|x| x.1.hir_id == *hir_id).map(|(i,_)| init_tup.get(i).unwrap()).next()
+                            };
+
+                            match init.kind {
+                                ExprKind::Tup(init_tup) => {
+                                    if let Some(init) = find_init(init_tup) {
+                                        return self.note_type_is_not_clone_inner_expr(init);
+                                    };
+                                    return None;
+                                }
+                                ExprKind::Block(
+                                    hir::Block {
+                                        expr: block_expr,
+                                        ..
+                                    },
+                                    _,
+                                ) => {
+                                    if let Some(Expr { kind: ExprKind::Tup(init_tup), .. }) = block_expr
+                                        && let Some(init) = find_init(init_tup)
+                                    {
+                                        return self.note_type_is_not_clone_inner_expr(init);
+                                    }
+                                    return None;
+                                }
+                                _ => {}
+                            }
+                        }
```

Comment:
this could use [`peel_blocks`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_hir/struct.Expr.html#method.peel_blocks)